File: ch01/ex1-3.R
################################################################################
# Usage: shell:
#   RScript ex1-3.R
# Usage: REPL:
#   r
#   > source('ex1-3.R')
################################################################################

square <- function(a) {
  a*a
}

# square(3) => 9

min <- function(a, b) {
  ifelse(a < b, a, b)
}

# min(2,3) => 2

min3 <- function(a, b, c) {
  min(a, min(b, c))
}

# min3(10, 8, 6) => 6

sumofsquares <- function(a, b, c) {
  square(a) + square(b) + square(c)
}

#sumofsquares(1, 2, 3)

q3 <- function(a, b, c) {
  sumofsquares(a, b, c) - square(min3(a, b, c))
}

q3(2, 3, 4)
q3(4, 3, 2)
q3(3, 1, 4)

File: ch01/ex1-3.scala
// scala ex1-3.scala

import scala.collection.mutable;

  object HelloWorld {

    def square(a:Int) : Int = {
      return a*a
    }

    def min(a:Int, b:Int) : Int = {
      if (a < b) {
        return a
      }
      return b
    }

    def min3(a:Int, b:Int, c:Int) : Int = {
      min(a, min(b, c))
    }

    def sumofsquares(a:Int, b:Int, c:Int) : Int = {
      square(a) + square(b) + square(c)
    }

    def q3(a:Int, b:Int, c:Int) : Int = {
      sumofsquares(a, b, c) - square(min3(a, b, c))
    }

    def main(args: Array[String]) {
      //println(square(5));
      //println(min(2,3));
      //println(min(3,2));
      //println(min3(2,3,4));
      //println(min3(3,4,2));
      //println(min3(4,2,3));
      //println(sumofsquares(2,3,4));
      println(q3(2,3,4));

    } // main
  }
